linode/ansible_linode
# ref: PG affinity_type, is_strict -> placement_group_type, placement_group_policy

## 📝 Description

This pull request implements the following upcoming breaking changes for the VM Placement GA:

* affinity_type -> placement_group_type
* is_strict -> placement_group_policy

**NOTE: We will need communication around this breaking change in the release notes.**

## ✔️ How to Test

The following test steps assume you have pulled down this PR locally and have pointed your local environment at an API instance where PGs are available:

```
export TEST_API_URL=https://.../
export TEST_API_CA=$PWD/cacert.pem
export LINODE_TOKEN=...

# Exports for dx-devenv sandbox
export LINODE_API_URL=$TEST_API_URL
export LINODE_CA=$TEST_API_CA
export LINODE_API_TOKEN=$LINODE_TOKEN
```

### Integration Testing

```bash
make TEST_ARGS="-v placement_group_assign placement_group_basic placement_group_info placement_group_list" test

# NOTE: This test is expected to fail in certain environments due to an unrelated issue, see TPT-3060 for more info.
make TEST_ARGS="-v instance_basic" test
```

### Manual Testing

1. In an ansible_linode sandbox environment (e.g. dx-devenv), run the following:

```yaml
- name: ansible_linode Sandbox Playbook
  hosts: localhost
  tasks:
    - name: Create a Linode placement group
      linode.cloud.placement_group:
        label: my-ansible-test-pg
        region: us-east
        placement_group_type: anti_affinity:local
        placement_group_policy: flexible
        state: present
      register: pg_created

    - name: Get a Linode placement group info
      linode.cloud.placement_group_info:
        id: '{{ pg_created.placement_group.id }}'
      register: pg

    - name: Update a Linode placement group label
      linode.cloud.placement_group:
        id: '{{ pg.placement_group.id }}'
        label: '{{ pg.placement_group.label }}-updated'
        placement_group_type: anti_affinity:local
        placement_group_policy: flexible
        state: present
      register: pg_updated

    - name: List Linode placement groups for the current account
      linode.cloud.placement_group_list:
        filters:
          - name: label
            values: ['{{ pg_updated.placement_group.label }}']
      register: pg_list
```

2. Observe no error is raised.

3. Create an instance under the placement group
```yaml
    - name: Create a Linode instance under a placement group
      linode.cloud.instance:
        label: 'ansible-test-pg-{{ r }}'
        type: g6-nanode-1
        region: us-east
        placement_group:
          id: '{{ pg_updated.placement_group.id }}'
          compliant_only: false
        state: present
      register: instance_pg
```

4. Observe that the instance is created successfully.

5. Clean up the resources created.
```yaml
    - name: Delete a Linode instance
      linode.cloud.instance:
        label: '{{ instance_pg.instance.label }}'
        state: absent
      register: delete_instance_pg

    - name: Delete a placement group
      linode.cloud.placement_group:
        label: '{{ pg_updated.placement_group.label }}'
        state: absent
      register: pg_deleted
```
